docs/smart-contracts/best-practices.md

```diff
@@ -0,0 +1,300 @@
+
+```pact
+(namespace 'free)
+(define-keyset "free.hello-world" (read-keyset 'ks))
+
+(module YODA5 GOV
+  (defcap GOV () 
+    (enforce-keyset "free.hello-world")) 
+  
+  (defcap USER (account:string guard:guard) 
+    (enforce-guard guard)) ; This condition requires the guard to be provided as input to sign for the transaction
+
+  (defun hello-world:string (input:string account:string guard:guard)
+    (with-capability (USER account guard)
+    (format "Hello {}" [input])))
+)
+```
+
+In this example, the `hello-world` function has been modified to require an `account` string and a `guard` as arguments.
+If the guard is a keyset, you can use the `read-keyset` function and keyset name to input the keys and predicate for the account.
+For example, if the guard is the keyset you defined using `ks` as the keyset name, you could call the function with arguments similar to the following:
+
+```pact
+(free.YODA5.hello-world "Robot" "k:4fe7981d36997c2a327d0d3ce961d3ae0b2d38185ac5e5cd98ad90140bc284d0" (read-keyset 'ks)) 
+"Hello Robot"
+```
+
+The following example illustrates another common enforcement mistake;
+
+```pact
+(namespace 'free)
+(define-keyset "free.hello-world" (read-keyset 'ks))
+
+(module YODA6 GOV
+  (defcap GOV () 
+    (enforce-keyset "free.hello-world"))
+
+  (defcap USER (account:string) 
+    (enforce (!= account "") "Specify an account")) ; Anyone can sign for this capability
+
+  (defun hello-world:string (input:string account:string)
+    (with-capability (USER account)
+    (format "Hello {}" [input])))
+)
+```
+
+With this pattern, any string used for the `account` argument passes the enforcement rule for the USER capability, enabling any user to acquire the capability and use the unprotected function.
+For example, any value can be used with this function:
+
+```pact
+(free.YODA6.hello-world "Robot" "jae") 
+"Hello Robot"
+```
+
+The following example illustrates using a hard-coded account string instead of reading a keyset or guard from a table or the message payload:
+
+```pact
+(namespace "free")
+(define-keyset "free.ks" (read-keyset 'ks))
+
+(module YODA7 GOV
+  (defcap GOV () 
+    (enforce-keyset "free.ks")) 
+  
+  (defcap USER (account:string) 
+    (enforce (= account "k:4fe7981d36997c2a327d0d3ce961d3ae0b2d38185ac5e5cd98ad90140bc284d0") "Invalid account"))
+  
+  (defun hello-world:string (input:string account:string)
+    (with-capability (USER account)
+    (format "Hello {}" [input])))
+)
+```
+
+If the enforced `account` string is used, the capability is acquired:
+
+```pact
+(free.YODA7.hello-world "Robot" "k:4fe7981d36997c2a327d0d3ce961d3ae0b2d38185ac5e5cd98ad90140bc284d0")
+"Hello Robot"
+```
+
+If any other account or arbitrary string is used, the operation fails:
+
+```pact
+(free.YODA7.hello-world "Robot" "k:9a23bf6a61f753d3ffa45c02b33c65b9dc80b8fb63857debcfe21fdb170fcd99")
+Error: <interactive>:9:4: Invalid account
+```
```

Comment:
I think this section really needs to be reworked from the start. As a beginner, this tells me nothing I care about.

Reply:
@edmundnoble I'm not sure what you have in mind. Feel free to elaborate or suggest a different approach. However, I don't think this page is, strictly-speaking, targeted for beginners. It assumes a certain level of familiarity with Pact code. I was thinking of it more as a summary of "simplified" reminders and examples of do's and don'ts about stuff covered elsewhere (like, in the capabilities and modules sections).